File: restChat.cpp
//
//  namesAPI.cc - a microservice demo program
//
// James Skon
// Kenyon College, 2022
//

#include <iostream>
#include <fstream>
#include <map>
#include <algorithm>
#include "httplib.h"

using namespace httplib;
using namespace std;

const int port = 5005;

void addMessage(string username, string message, map<string,vector<string>> &messageMap) {
	/* iterate through users adding message to each */
	string jsonMessage = "{\"user\":\""+username+"\",\"message\":\""+message+"\"}";
	for (auto userMessagePair : messageMap) {
		username = userMessagePair.first;
		messageMap[username].push_back(jsonMessage);
	}
}

// Return messages for a certain user
string getMessagesJSON(string username, map<string,vector<string>> &messageMap) {
	/* retrieve json list of messages for this user */
	bool first = true;
	string result = "{\"messages\":[";
	for (string message :  messageMap[username]) {
		if (not first) result += ",";
		result += message;
		first = false;
	}
	result += "]}";
	messageMap[username].clear();
	return result;
}

// My function to get list of users
void addUsers(string username, map<string,vector<string>> &messageMap, vector<string> userList) {
	// Create a json list of current users
	string jsonUser = "{\"user\":\""+username+"\"}";
	for (auto userMessagePair : messageMap) {
		username = userMessagePair.first;
		userList[username].push_back(jsonUser);
	}
}

// Return a json list of user 
string getUsersJSON(string username, vector<string> &userList) {
	/* retrieve json list of users */
	bool first = true;
	string result = "{\"user\":[";
	for (string user :  userList[username]) {
		if (not first) result += ",";
		result += user;
		first = false;
	}
	result += "]}";
	userList[username].clear();
	return result;
}

int main(void) {
  Server svr;
  int nextUser=0;
  map<string,vector<string>> messageMap;
  vector<string> userList;
	
  /* "/" just returnsAPI name */
  svr.Get("/", [](const Request & /*req*/, Response &res) {
    res.set_header("Access-Control-Allow-Origin","*");
    res.set_content("Chat API", "text/plain");
  });


  svr.Get(R"(/chat/join/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
    string username = req.matches[1];
    string resultMessage;
    string resultUser;
    vector<string> empty;
    cout << username << " joins" << endl;
    
    // Check if user with this name exists
    if (messageMap.count(username)) {
    	resultUser = "{\"status\":\"exists\"}";
    } else {
    	// Add user to messages map
    	messageMap[username]=empty;
    	resultMessage = "{\"status\":\"success\",\"user\":\"" + username + "\"}";
    	// Add user to user list
    	userList[username]=empty;
    	resultUser = "{\"status\":\"success\",\"user\":\"" + username + "\"}";
    }
    res.set_content(resultMessage, "text/json");
    res.set_content(resultUser, "text/json");
  });

   svr.Get(R"(/chat/send/(.*)/(.*))", [&](const Request& req, Response& res) {
    res.set_header("Access-Control-Allow-Origin","*");
	string username = req.matches[1];
	string message = req.matches[2];
	string resultMessage; 
	string resultUser; 
	
    if (!messageMap.count(username)) {
    	resultUser = "{\"status\":\"baduser\"}";
	} else {
		addMessage(username,message,messageMap);
		resultMessage = "{\"status\":\"success\"}";
		addUsers(username,messageMap,userList);
		resultUser = "{\"status\":\"success\"}";
	}
    res.set_content(resultMessage, "text/json");
    res.set_content(resultUser, "text/json");
  });
  
   svr.Get(R"(/chat/fetch/(.*))", [&](const Request& req, Response& res) {
    string username = req.matches[1];
    res.set_header("Access-Control-Allow-Origin","*");
    string resultMessageJSON = getMessagesJSON(username,messageMap);
    res.set_content(resultMessageJSON, "text/json");
    string resultUsersJSON = getUsersJSON(username,userList);
    res.set_content(resultUsersJSON, "text/json");
  });
  
  cout << "Server listening on port " << port << endl;
  svr.listen("0.0.0.0", port);

}

File: restChat.js
// Rest based chat client
// Jim Skon 2022
// Kenyon College

var baseUrl = 'http://3.15.139.27:5005';
var state="off";
var myname="";
var inthandle;
var inthandleUsers;
var currentUsers = [];
var masterUsers = [];

/* Start with text input and status hidden */
document.getElementById('chatinput').style.display = 'none';
document.getElementById('status').style.display = 'none';
document.getElementById('leave').style.display = 'none';

/* Set up buttons */
// Action if they push the join button
document.getElementById('login-btn').addEventListener("click", (e) => {
	join();
})
// Action if they push the leave button
document.getElementById('leave-btn').addEventListener("click", (e) => {
	leaveSession();
})
document.getElementById('send-btn').addEventListener("click", sendText);
// Action if they push enter on message box
document.getElementById('message').addEventListener("keydown", (e)=> {
    if (e.code == "Enter") {
	e.preventDefault();
    	sendText();
    	document.getElementById('message').value = "";
    	return false;
    }   
});

// Call function on page exit
window.onbeforeunload = leaveSession;

/* Join processes */
function join() {
	myname = document.getElementById('yourname').value;
	fetch(baseUrl+'/chat/join/'+myname, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data => completeJoin(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })
}

function completeJoin(results) {
	var user = results['user'];
	if (currentUsers.includes(user) == true){
		alert("Username already exists!");
		leaveSession();
		return;
	}
	console.log("Join:"+user);
	startSession(user);
	logUser(user);
}

/* Log users (only on client side right now) */
function logUser(user) {
	if (currentUsers.includes(user) == false) {
		currentUsers.push(user);
	};
	for (var i = 0; i < currentUsers.length; i++) {
  		document.getElementById("members").innerHTML += (i+1) + ": " + currentUsers[i] + " ";
	};
}

function removeUser(user) {
	for (var i = 0; i < currentUsers.length; i++) {
		if (user == currentUsers[i]){
			currentUsers.splice(i,1);
		}
	};
	document.getElementById("members").innerHTML = " ";
	for (var i = 0; i < currentUsers.length; i++) {
  		document.getElementById("members").innerHTML += (i+1) + ": " + currentUsers[i] + " ";
	};
}

/* Register Users */
function registerUser() {
	username = document.getElementById('register-name');
	email = document.getElementById('register-email');
	password = document.getElementById('register-password');
	fetch(baseUrl+'/chat/join/'+username+'/'+email+'/'+password, {
		method: 'get'
	})
	.then (response => response.json())
	.then (data => completeRegisterUser(data))
	.catch(error => {
		{alert("Error: Something went wrong:"+error);}
	})
}

function completeRegisterUser(results) {
	var status = results['status'];
	if (status != "success") {
		alert("Username or email already exists!");
		leaveSession();
		return;
	}
	var user = results['user'];
	console.log("Register:"+user);
	startSession(user);
	logUsers(user);
}

function completeSend(results) {
	var status = results['status'];
	if (status == "success") {
		console.log("Send succeeded")
	} else {
		alert("Error sending message!");
	}
}

//function called on submit or enter on text input
function sendText() {
    var message = document.getElementById('message').value;
    console.log("Send: "+myname+":"+message);
	fetch(baseUrl+'/chat/send/'+myname+'/'+message, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data => completeSend(data))
    .catch(error => {
        {alert("Error: Something went wrong:"+error);}
    })    
	document.getElementById('message').value = "";
}

function completeFetch(result) {
	// Messages
	messages = result["messages"];
	messages.forEach(function (m,i) {
		name = m['user'];
		message = m['message'];
		document.getElementById('chatBox').innerHTML +=
	    	"<font color='red'>" + name + ": </font>" + message + "<br />";
	});
	// Users
	users = result["userList"];
	users.forEach(function (m,i) {
		name = m['user'] + ", ";
		if (masterUsers.includes(name) == false) {
			masterUsers.push(name);
		}
	});
}

/* Check for new messaged */
function fetchMessage() {
	fetch(baseUrl+'/chat/fetch/'+myname, {
        method: 'get'
    })
    .then (response => response.json() )
    .then (data =>completeFetch(data))
    .catch(error => {
        {console.log("Server appears down");}
    })  	
}

/* Functions to set up visibility of sections of the display */
function startSession(name){
    state="on";
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'none';
    document.getElementById('user').innerHTML = "User: " + name;
    document.getElementById('chatinput').style.display = 'block';
    document.getElementById('status').style.display = 'block';
    document.getElementById('leave').style.display = 'block';
    /* Check for messages every 500 ms */
    inthandle=setInterval(fetchMessage,500);
    /* Check for current users every 500 ms */
    //inthandleUsers=setInterval(updateUsers,500);
}

function leaveSession(){
    state="off";
    logout();
    document.getElementById('yourname').value = "";
    document.getElementById('register').style.display = 'block';
    document.getElementById('user').innerHTML = "";
    document.getElementById('chatinput').style.display = 'none';
    document.getElementById('members').style.display = 'none';
    document.getElementById('status').style.display = 'none';
    document.getElementById('leave').style.display = 'none';
	clearInterval(inthandle);
	//clearInterval(inthandleUsers);
}

function logout() {
	fetch(baseUrl+'/chat/logout/'+myname, {
		method: 'get'
	})
	.then (response => response.json() )
	.then (data => completeLogout(data) )
	.catch(error => {
		{console.log("Service Unavailable");}
	})
}

function completeLogout(user) {
	removeUser(user);
}


